### ISSUE TITLE
Image with non-nearest, non-linear interpolation and `visible=False` => WEIRD

### ISSUE THREAD

**Issue by brisvag** [napari/napari#6485]
### 🐛 Bug Report

Weird title for a weird bug. My plugin just happens to generate layers like this. As far as I can tell, these are the minimal conditions for the bug to happen, but there's a few combinations that cause it.

Whe the bug appears, it seems to be unfixable, and simply shows a big blurry blob instead of the actual image (note the thumbnail mismatch):

![image](https://github.com/napari/napari/assets/23482191/bfde92eb-a7c1-465a-b775-b838383fa24e)

### 💡 Steps to Reproduce

```py
import napari
import numpy as np

v = napari.Viewer()
a = np.random.rand(10, 10)
il = v.add_image(a, visible=False, interpolation2d='spline36')
```

Now, if you make the layer visible via the GUI button or via `il.visible = True`, all will be fine.

However, if you press the `v` shortcut to make it visible, you'll see the bug (and won't be able to fix it).

This is consistent.

However, now introduce `dask`:

```py
import napari
import numpy as np
import dask.array as da

v = napari.Viewer()
a = np.random.rand(10, 10, 10)
d = da.array(a)
il = v.add_image(d, visible=False, interpolation2d='spline36')
```

Suddenly, the bug shows up even when you press the gui button or use the programmatic interface, but only *sometimes*.

I'm not sure on this, but I have the impression that adding `NAPARI_ASYNC=1` to the mix makes the bug appear even more often... I guess this points to some sort of race condition?

### 💡 Expected Behavior

Normal image shows up :P

### 🌎 Environment

napari main

**Comment by brisvag** [comment]
*might* be related to clim? @melonora @Czaki

**Comment by brisvag** [comment]
Since you couldn't reproduce, I retested in a new env and the issue is not happening anymore.

So I slowly updated packages in the old env until `napari --info` returns exactly the same output, and the old env is *still* broken...

**Comment by psobolewskiPhD** [comment]
I ran both snippets 2 or 3 times and toggled vis with `v` buncha times and never saw anything wierd.
So it's quite strange!

**Comment by brisvag** [comment]
Actually, now the new env is also broken.... This is super weird!

**Comment by brisvag** [comment]
Yup, confirm that new, fresh environment is also broken. I guess I was just "lucky" a couple times in a row earlier and didn't see the issue. Will dig a bit deeper now.

**Comment by psobolewskiPhD** [comment]
Do you have a repo for the plugin to `pip install git+....` ?

**Comment by brisvag** [comment]
@andy-sweet might have insights here.

**Comment by brisvag** [comment]
This line was added in that PR and looks suspicious:

https://github.com/napari/napari/blob/85c58b5990dfbbc41468ae16c25216e8a00f11a3/napari/components/_layer_slicer.py#L214

**Comment by brisvag** [comment]
Yup, that's the line :) If I replace `visible_layers` back with `layers`, the bug is gone. Why this is irreproducible, I don't know... but it seems to be inconsistent even on the same machine and env, which means there's some randomness to it, maybe to deal with data transfer speed to the GPU?

**Comment by psobolewskiPhD** [comment]
OK with NAPARI_ASYNC=1 I get the issue with both snippets using `v` every time I've tried (5 times) but not using the eye button (4 times). 👀 
I can recover though by changing the interpolation to nearest and then back to spline.

**Comment by andy-sweet** [comment]
I can take a look and try to provide an explanation. Thanks for all the info!

**Comment by andy-sweet** [comment]
FWIW, I don't see differences between using the `v` keyboard shortcut, the eye, or the API on either example with async on or off.

But I can reliably reproduce the issue with async off and have not been able to reproduce it with async off.

All the state on the layer and the vispy node looks like I'd expect, so I'm left to assume this is an ordering issue.

> I can recover though by changing the interpolation to nearest and then back to spline.

Changing the interpolation was the only workaround I found too. So I think this could be pretty important to understanding the bug.

In terms of potentially important differences between sync and async in this specific case (i.e. when the `visible` parameter is `False` on initialization), I can't think of too many because in both cases slicing should be initially skipped. 

When `visible` changes to `True`, the layer will explicitly refresh itself, then emit its `visible` event, which will then trigger its callbacks. By default `VispyCanvas._reorder_layers` and `VispyImageLayer._on_visible_change` are these callbacks. The more interesting one here is `_reorder_layers` because that calls `VispyImageLayer._on_blending_change` (which does a few interesting things).

For sync slicing, `refresh` executes, which will eventually execute the `VispyImageLayer._on_data_change`, then `_reorder_layers` is executed, which calls `VispyImageLayer._on_blending_change`.

For async slicing, `refresh` will cause a slicing job/request to be submitted, then `_reorder_layers` is executed, then sometime later `VispyImageLayer._on_data_change` is called.

Unfortunately, calling `_on_blending_change` or `_reorder_layers` explicitly after changing `visible` does not fix things for `async`, so this isn't a full explanation. But maybe it's still relevant?

Will try to do some more digging later.

**Comment by andy-sweet** [comment]
Two other hopefully useful tidbits for vispy experts.

1\. I cannot reproduce this bug when using `nearest` or `linear` interpolation. But can for all the others.

2\. A minimal way to workaround this (after `visible` changes to `True`), is to set `VispyImageLayer.node._need_interpolation_update = True` then call `VispyImageLayer.node.update()`.

Are the non-nearest/linear interpolation methods dependent on the vispy node's data in some way? If that's true then it could be that the ordering above + the observations in this comment explains the problem (because after `visible` changes to `True`, `_on_blending_change` sets `_need_interpolation_update = False` before the node data has been set).

**Comment by jni** [comment]
I have nothing to add here, but it was a fun read. 😂

**Comment by brisvag** [comment]
Thanks for digging @andy-sweet.

> 1. I cannot reproduce this bug when using `nearest` or `linear` interpolation. But can for all the others.

> Are the non-nearest/linear interpolation methods dependent on the vispy node's data in some way? If that's true then it could be that the ordering above + the observations in this comment explains the problem (because after `visible` changes to `True`, `_on_blending_change` sets `_need_interpolation_update = False` before the node data has been set).

I saw the same; these two interpolations are hardware based (they just change the way the texture is sampled), while the rest are shaders (so multiple texture samples are combined by some code). However, I don't know how this might be related to the data itself...
